lib_test/test_httpaf.ml

```diff
@@ -128,10 +128,54 @@ module Server_connection = struct
     ;;
   end
 
+  module Write_operation = struct
+    type t = [ `Write of Bigstringaf.t IOVec.t list | `Yield | `Close of int ]
+
+    let iovecs_to_string iovecs =
+      let len = IOVec.lengthv iovecs in
+      let bytes = Bytes.create len in
+      let dst_off = ref 0 in
+      List.iter (fun { IOVec.buffer; off = src_off; len } ->
+        Bigstringaf.unsafe_blit_to_bytes buffer ~src_off bytes ~dst_off:!dst_off ~len;
+        dst_off := !dst_off + len)
+      iovecs;
+      Bytes.unsafe_to_string bytes
+    ;;
+
+    let pp_hum fmt t =
+      match t with
+      | `Write iovecs -> Format.fprintf fmt "Write %S" (iovecs_to_string iovecs)
+      | `Yield -> Format.pp_print_string fmt "Yield"
+      | `Close len -> Format.fprintf fmt "Close %i" len
+    ;;
+
+    let to_write_as_string t =
+      match t with
+      | `Write iovecs -> Some (iovecs_to_string iovecs)
+      | `Close _ | `Yield -> None
+    ;;
+  end
+
+  let read_string t str =
+    let len = String.length str in
+    let input = Bigstringaf.of_string str ~off:0 ~len in
+    read t input ~off:0 ~len;
+  ;;
+
   let default_request_handler reqd =
     Reqd.respond_with_string reqd (Response.create `OK) ""
   ;;
 
+  let synchronous_raise reqd =
+    Reqd.report_exn reqd (Failure "caught this exception")
```

Comment:
Just for clarification: we are calling this a synchronous "raise", but it's not actually raising -- would we ever test the case where an exception is actually raised, or are expecting the implementations to wrap user functions and catch all exceptions?

Reply:
Correct, it's not raising. An uncaught exception should make its way back to the state machine via the runtime, which I imagine isn't currently happening. To raise in a test without a runtime would just cause the test program to crash.

Reply:
The Lwt runtime does some form of exception reporting, e.g.:
https://github.com/inhabitedtype/httpaf/blob/8ed4fe131d0fd39b8b42e6d5eb2dc10209d752b2/lwt/httpaf_lwt.ml#L124

The Async runtime doesn't which is what #95 does in one way too.

Reply:
One thing that I've been thinking about that probably needs to be fixed in the runtimes is to only report the exception back to the state machine if the socket is not closed, otherwise it's probably pointless to do so given it'll most likely result in the runtime raising an `EPIPE` error.